Entrega: brute-force 8-digit SHA-256 (threaded) — senhas encontradas

djoseph-png opened the pull request:
Resumo
------
Implementação para a tarefa de brute-force de 10 hashes SHA-256 (senhas numéricas de 8 dígitos).

O que fiz
---------
- Substituí/adicionei `app/main.py` (versão threaded) que executa a busca em paralelo usando threads.
- Adicionei `found_passwords.txt` (10 senhas, uma por linha).
- Adicionei `found_output.txt` (log completo do terminal) e `screenshot_terminal.png` com a evidência visual.

Como rodar
----------
1. Ative o venv:
   .\venv\Scripts\Activate.ps1

2. Execute:
   python app/main.py -p 10

Observações técnicas
-------------------
- Usei threads porque `hashlib.sha256()` é implementado em C e libera o GIL, o que permite que threads alcancem alto throughput sem o overhead do multiprocessing no Windows.
- Em minha máquina (Python 3.13, Windows), o tempo total para varrer o espaço — na configuração `-p 10` — foi ~137s (como mostrado no log). As senhas encontradas estão em `found_passwords.txt`.
- Se preferirem a versão `multiprocessing`, mantenho o histórico e posso remover/alterar conforme o requisito do avaliador.

Arquivos incluídos
------------------
- `app/main.py` — script de execução
- `found_passwords.txt` — as 10 senhas encontradas
- `found_output.txt` — log do console (opcional)
- `screenshot_terminal.png` — evidência visual do terminal

Notas finais
-----------
- Não incluí `venv` no commit (.gitignore deve ignorá-lo). Se precisar que rode em outro ambiente, indico a versão do Python (3.11/3.13) e as dependências (nenhuma extra — stdlib).
<img width="563" height="300" alt="screenshot_terminal" src="https://github.com/user-attachments/assets/c53cc256-d73d-4c18-aacd-6a93986eff0f" />

djoseph-png committed:
Entrega: implementação threaded brute-force + relatório. Impressões: 10 senhas encontradas.

djoseph-png committed:
Adiciona screenshot do terminal com 10 senhas

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -16,17 +23,147 @@
+    flush_every: int = 2000,
+    check_done_every: int = 2000,
+) -> None:
+    _sha256 = sha256
```
The task description requires the helper function `sha256_hash_str(to_hash: str) -> str` which must perform `sha256(to_hash.encode("utf-8")).hexdigest()`. Currently the code computes the digest inline rather than providing/using this function. Add the required function and call it from the worker to match the explicit checklist item in the description.

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -16,17 +23,147 @@
+
+    for n in range(r.start, r.end):
+        candidate = f"{n:08d}"
+        digest = _sha256(_encode(candidate, "ascii")).hexdigest()
```
This line uses ASCII encoding when building the digest: `digest = _sha256(_encode(candidate, "ascii")).hexdigest()`. The specification explicitly requires UTF-8 encoding (`to_hash.encode("utf-8")`). Change the code to use UTF-8 (ideally by calling `sha256_hash_str(candidate)`) to follow the exact requirement. Even though digits are identical in ASCII and UTF-8, the task is explicit about the encoding and the helper function.

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -16,17 +23,147 @@
+    check_done_every: int = 2000,
+) -> None:
+    _sha256 = sha256
+    _encode = str.encode
```
Instead of binding `_sha256` and `_encode` and hashing inline, centralize hashing by implementing `sha256_hash_str` and call it here: `digest = sha256_hash_str(candidate)`. This satisfies the required API and improves clarity/testability.

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -16,17 +23,147 @@
+    check_counter = 0
+
+    for n in range(r.start, r.end):
+        candidate = f"{n:08d}"
```
Candidate generation is correct and meets the requirement to produce all 8-digit zero-padded numeric strings: `candidate = f"{n:08d}"`. This preserves leading zeros as required by the task.

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -16,17 +23,147 @@
+
+
+def brute_force_threads(processes: int | None = None) -> list[str]:
+    targets = set(PASSWORDS_TO_BRUTE_FORCE)
```
Targets are stored in a set (`targets = set(PASSWORDS_TO_BRUTE_FORCE)`), which is the expected efficient data structure for membership tests per the checklist. Keep this usage.

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -16,17 +23,147 @@
+                    counter[0] += local_count
+                local_count = 0
+            with results_lock:
+                if digest not in results:
+                    results[digest] = candidate
+                    print(f"FOUND: {candidate} -> {digest}", flush=True)
+                    if len(results) == len(_targets):
+                        done.set()
+                        break
```
Result mapping and early termination are implemented correctly: when a digest match is found it's added to `results` and when `len(results) == len(_targets)` the `done` event is set to stop other workers. This satisfies the requirement to detect completion when all 10 passwords are found and to avoid duplicates.

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -16,17 +23,147 @@
+    found_passwords.sort(key=lambda s: int(s, 10))
+    print(f"# Tempo total: {dt:.2f}s  encontrados: {len(found_passwords)}/{len(targets)}", flush=True)
+    return found_passwords
 
 
 if __name__ == "__main__":
-    start_time = time.perf_counter()
-    brute_force_password()
-    end_time = time.perf_counter()
+    import argparse
+
+    parser = argparse.ArgumentParser(description="Threaded brute-force 8-digit numeric passwords (SHA-256).")
+    parser.add_argument("-p", "--processes", type=int, default=None, help="número de threads (padrão = qtd. de CPUs).")
+    args = parser.parse_args()
 
-    print("Elapsed:", end_time - start_time)
+    pwds = brute_force_threads(args.processes)
+    for pwd in pwds:
+        print(pwd)
```
The code collects, sorts and prints found passwords (`found_passwords` and later printing in main). This fulfils the requirement to output plaintext passwords to the console for verification. Make sure after adding `sha256_hash_str` the printed outputs still include all 10 discovered passwords.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

Your implementation is close, but there’s a critical checklist violation that blocks approval.

- Critical blocker (must fix): The task explicitly requires a helper `sha256_hash_str(to_hash: str) -> str` that returns `sha256(to_hash.encode("utf-8")).hexdigest()`. In `app/main.py`, you compute hashes inline using ASCII: `digest = _sha256(_encode(candidate, "ascii")).hexdigest()` (see worker thread; please reference the exact line in your file). This violates the required API and encoding. Even though ASCII and UTF-8 are identical for digits, the checklist mandates UTF-8 and the named helper.
- How to fix: Implement the helper exactly as specified and replace inline hashing with `digest = sha256_hash_str(candidate)`. This both satisfies the explicit requirement and centralizes hashing for clarity/testability.

What’s good: Candidate generation with `f"{n:08d}"`, using a `set` for `targets`, the results mapping, and the early termination logic all meet the requirements, and printing the found passwords is correctly implemented. After adding the helper and using UTF-8 via that function, your submission should fully comply. Keep up the solid work—this is a quick adjustment to get across the finish line.

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

djoseph-png committed:
Fix: add sha256_hash_str(to_hash: str) using utf-8 and replace inline ascii hashing; threaded brute-force finds 10 passwords.

djoseph-png commented:
Fix: implement sha256_hash_str(to_hash: str) -> str and replace inline hashing

- Implemented required helper: sha256_hash_str(to_hash: str) -> str which returns
  sha256(to_hash.encode("utf-8")).hexdigest().
- Replaced inline hashing calls (previously using ASCII encoding) with
  digest = sha256_hash_str(candidate) in worker_thread to meet the checklist.
- Kept optimized threaded brute-force implementation for performance.
- Added found_passwords.txt and found_output.txt with evidence (10/10).

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -16,17 +24,163 @@
+    procs = processes or cpu_count
+    total = 100_000_000
+
+    results: dict = {}
```
Consider adding a precise type annotation to the `results` mapping to make the types explicit for readers and linters. For example: `results: dict[str, str] = {}`. This is optional (not required by the checklist) but improves clarity.

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -16,17 +24,163 @@
+    targets = set(PASSWORDS_TO_BRUTE_FORCE)
+    cpu_count = os.cpu_count() or 1
+    procs = processes or cpu_count
+    total = 100_000_000
```
The `total` variable is correct (`100_000_000`) and represents the full 8-digit numeric space. As a minor readability suggestion you could write it as `total = 10**8` to make the intent explicit; no functional change required.

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -16,17 +24,163 @@
+    args = parser.parse_args()
 
-    print("Elapsed:", end_time - start_time)
+    pwds = brute_force_threads(args.processes)
+    for pwd in pwds:
+        print(pwd)
```
The task requires attaching a screenshot of the script results (console output) to the PR. The script prints the discovered passwords in the main block — make sure to include a screenshot of that output when you submit the PR.

mateacademy-ai-mentor reviewed (approved):
### Review Summary

Great work — the implementation in app/main.py meets the task requirements. You correctly implemented `sha256_hash_str(to_hash: str) -> str` using UTF-8 and `hexdigest()`, generated the exact 8-digit numeric candidate space (`f"{n:08d}"`), used a set for target lookups, maintained a thread-safe results map with early termination when all 10 are found, and printed the discovered passwords in the main block.

> Reminder: The task also asks for proof of results. Please attach a screenshot of the console output that shows the found passwords when submitting the PR.

Minor, non-blocking improvements:
- Add a precise type annotation for `results` (e.g., `results: dict[str, str] = {}`) for clarity.
- Optionally express the search space as `total = 10**8` for readability.

Overall, this is solid and ready to merge once you include the screenshot. Nicely done!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)
